File: fabric/src/main/java/net/shiroha233/roadweaver/planning/fabric/ServerPlanningHooks.java
/* 文件职责：注册 Fabric 服务端生命周期与 Tick 钩子。 */
package net.shiroha233.roadweaver.planning.fabric;

import net.fabricmc.fabric.api.event.lifecycle.v1.ServerLifecycleEvents;
import net.fabricmc.fabric.api.event.lifecycle.v1.ServerTickEvents;
import net.minecraft.server.level.ServerLevel;
import net.minecraft.server.level.ServerPlayer;
import net.minecraft.world.level.Level;
import net.shiroha233.roadweaver.config.structure.StructureDiscoveryService;
import net.shiroha233.roadweaver.core.model.StructureConnection;
import net.shiroha233.roadweaver.features.path.decoration.text.SignTextService;
import net.shiroha233.roadweaver.generation.IdleRoadGenerationService;
import net.shiroha233.roadweaver.generation.InitialGenManager;
import net.shiroha233.roadweaver.generation.RoadGenerationService;
import net.shiroha233.roadweaver.persistence.WorldDataProvider;
import net.shiroha233.roadweaver.persistence.sqlite.H2MigrationCoordinator;
import net.shiroha233.roadweaver.planning.RoadPlanningService;
import net.shiroha233.roadweaver.runtime.CacheManager;
import net.shiroha233.roadweaver.runtime.ThreadPoolManager;

import java.util.List;

/**
 * Fabric 服务端规划钩子
 */
public final class ServerPlanningHooks {
    private ServerPlanningHooks() {}

    private static int tick;

    public static void register() {
        ServerLifecycleEvents.SERVER_STARTED.register(server -> {
            H2MigrationCoordinator.migrateServer(server);
            CacheManager.onServerStarted();
            ThreadPoolManager.onServerStarted(server);
            SignTextService.clearPending();
            IdleRoadGenerationService.onServerStarted();
            
            ServerLevel level = server.getLevel(net.minecraft.world.level.Level.OVERWORLD);
            if (level == null) return;

            StructureDiscoveryService.discoverFromLevel(level);

            boolean dedicated = server.isDedicatedServer();
            if (dedicated) {
                RoadGenerationService.onServerStarted();
                if (InitialGenManager.shouldRunInitialGeneration(level)) {
                    RoadPlanningService.initialPlanAsync(level);
                }
                return;
            }
            
            if (InitialGenManager.shouldRunInitialGeneration(level)) {
                RoadGenerationService.onServerStarted();
                InitialGenManager.begin(level);
                InitialGenManager.blockUntilDone(level);
            } else {
                RoadGenerationService.onServerStarted();
            }
        });

        ServerTickEvents.END_SERVER_TICK.register(server -> {
            if ((tick++ % 20) == 0) {
                for (ServerPlayer p : server.getPlayerList().getPlayers()) {
                    if (Level.OVERWORLD.equals(p.serverLevel().dimension())) {
                        SignTextService.onChunkReady(p.serverLevel(), p.chunkPosition());
                        RoadPlanningService.planAroundPlayer(p);
                        IdleRoadGenerationService.tickPlayer(p);
                    }
                }
            }

            ServerLevel level = server.getLevel(Level.OVERWORLD);
            if (level != null) {
                IdleRoadGenerationService.tick(level);
                RoadGenerationService.tick(level);
                SignTextService.tick(level);
            }
        });

        ServerLifecycleEvents.SERVER_STOPPING.register(server -> {
            RoadGenerationService.onServerStopping();
            RoadPlanningService.resetAll();
            CacheManager.onServerStopping(server.getAllLevels());
            ThreadPoolManager.onServerStopping();
            SignTextService.clearPending();
        });
    }
}

File: fabric/src/main/kotlin/net/shiroha233/roadweaver/planning/fabric/ServerPlanningHooks.kt
package net.shiroha233.roadweaver.planning.fabric

import net.fabricmc.fabric.api.event.lifecycle.v1.ServerLifecycleEvents
import net.fabricmc.fabric.api.event.lifecycle.v1.ServerTickEvents
import net.minecraft.server.level.ServerLevel
import net.minecraft.server.level.ServerPlayer
import net.minecraft.world.level.Level
import net.shiroha233.roadweaver.generation.InitialGenManager
import net.shiroha233.roadweaver.generation.RoadGenerationService
import net.shiroha233.roadweaver.helpers.Records
import net.shiroha233.roadweaver.persistence.WorldDataProvider
import net.shiroha233.roadweaver.planning.RoadPlanningService
import net.shiroha233.roadweaver.runtime.CacheManager
import net.shiroha233.roadweaver.runtime.ThreadPoolManager
import net.shiroha233.roadweaver.util.ComputeService

object ServerPlanningHooks {
    private var tick: Int = 0

    @JvmStatic
    fun register() {
        ServerLifecycleEvents.SERVER_STARTED.register { server ->
            CacheManager.onServerStarted() // 统一初始化缓存
            ThreadPoolManager.onServerStarted(server)
            val level: ServerLevel = server.getLevel(Level.OVERWORLD) ?: return@register

            // 无论是新世界还是已存在的世界，都发现并缓存结构（供结构选择 GUI 使用）
            net.shiroha233.roadweaver.config.structure.StructureDiscoveryService.discoverFromLevel(level)

            val dedicated = server.isDedicatedServer
            if (dedicated) {
                RoadPlanningService.initialPlanAsync(level)
                return@register
            }

            val conns: List<Records.StructureConnection> = WorldDataProvider.getInstance().getStructureConnections(level)
            if (conns.isEmpty()) {
                InitialGenManager.begin(level)
                InitialGenManager.blockUntilDone(level)
            }
        }

        ServerTickEvents.END_SERVER_TICK.register { server ->
            if ((tick++ % 20) == 0) {
                for (p: ServerPlayer in server.playerList.players) {
                    RoadPlanningService.planAroundPlayer(p)
                }
            }
            val level: ServerLevel? = server.getLevel(Level.OVERWORLD)
            if (level != null) {
                RoadGenerationService.tick(level)
            }
        }

        ServerLifecycleEvents.SERVER_STOPPING.register { server ->
            RoadGenerationService.onServerStopping()
            CacheManager.onServerStopping(server.allLevels) // 统一清理所有缓存
            ComputeService.shutdownNow()
        }
    }
}
